src/main/java/com/epam/aidial/core/service/CacheService.java

```diff
@@ -0,0 +1,447 @@
+package com.epam.aidial.core.service;
+
+import com.epam.aidial.core.storage.BlobStorage;
+import com.epam.aidial.core.storage.BlobStorageUtil;
+import com.epam.aidial.core.storage.ResourceDescription;
+import com.epam.aidial.core.util.Compression;
+import com.epam.aidial.core.util.ResourceUtil;
+import com.google.common.collect.Sets;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import lombok.Builder;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.redisson.api.RMap;
+import org.redisson.api.RScoredSortedSet;
+import org.redisson.api.RedissonClient;
+import org.redisson.client.codec.StringCodec;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+
+@Slf4j
+public class CacheService implements AutoCloseable {
+    private static final String BODY_ATTRIBUTE = "body";
+    public static final String CONTENT_TYPE_ATTRIBUTE = "content_type";
+    public static final String CONTENT_LENGTH_ATTRIBUTE = "content_length";
+    private static final String BINARY_ATTRIBUTE = "binary";
+    private static final String SYNCED_ATTRIBUTE = "synced";
+    private static final String EXISTS_ATTRIBUTE = "exists";
+    private static final Set<String> REDIS_FIELDS_NO_BODY = Set.of(
+            ResourceUtil.ETAG_ATTRIBUTE,
+            ResourceUtil.CREATED_AT_ATTRIBUTE,
+            ResourceUtil.UPDATED_AT_ATTRIBUTE,
+            CONTENT_TYPE_ATTRIBUTE,
+            CONTENT_LENGTH_ATTRIBUTE,
+            BINARY_ATTRIBUTE,
+            SYNCED_ATTRIBUTE,
+            EXISTS_ATTRIBUTE);
+    private static final Set<String> REDIS_FIELDS = Sets.union(
+            Set.of(BODY_ATTRIBUTE),
+            REDIS_FIELDS_NO_BODY);
+
+    private static final Base64.Encoder BASE64_ENCODER = Base64.getEncoder();
```

Comment:
Please try using binary codec instead of base64 encoding.

Reply:
Done.